docs/proposals/THV-2301-refactor-mcpregistry-controller.md

```diff
@@ -0,0 +1,315 @@
+- Fixed 5-minute retry interval
+- Unlimited retry attempts
+- `syncStatus.attemptCount` increments on failure, resets on success
+
+**Proposed (Phase 1):**
+Match current behavior - Registry API retries every 5 minutes indefinitely.
+
+**Future enhancement:**
+Add configurable retry policy to `syncPolicy.retryPolicy` (tracked separately).
+
+### Data Filtering
+
+- Registry API applies name and tag filters during sync processing
+- Filter changes trigger immediate sync (hash-based detection)
+- Move `cmd/thv-operator/pkg/filtering` package to `thv-registry-api`
+
+### API Status Updates
+
+Controller continues managing `apiStatus` with no changes:
+- Monitors Deployment and Service health
+- Updates phase: NotStarted, Deploying, Ready, Unhealthy, Error
+- Existing `apiStatus` structure preserved
+
+## Implementation Plan
+
+### Phase 1: Move Data Source Management to Registry API
+
+**Goal:** Full-functioning solution with sync logic in `thv-registry-api`, maintaining backward compatibility with existing MCPRegistry resources.
+
+**Task Order Rationale:** Registry API must be enhanced first before operator changes, to avoid breaking existing deployments.
+
+#### Registry API Development
+
+1. **Add configuration file support**
+   - Implement `--config` flag in `thv-registry-api serve` command
+   - Add YAML configuration parsing for `source`, `syncPolicy`, and `filter` fields
+   - Implement file watcher for configuration changes (using `fsnotify` or similar)
+   - Graceful reload on configuration changes
+
+2. **Add file-based storage**
+   - Implement `--storage-path` flag for registry data location
+   - Add file-based storage backend matching current ConfigMap JSON format
+   - Atomic file writes (temp file + rename pattern)
+   - Load existing data on startup if present
+
+3. **Port source handler packages**
+   - Copy `cmd/thv-operator/pkg/sources` to `thv-registry-api`
+   - Copy `cmd/thv-operator/pkg/httpclient` to `thv-registry-api`
+   - Copy `cmd/thv-operator/pkg/git` to `thv-registry-api`
+   - Adapt packages to work without Kubernetes client dependencies
+   - Update import paths
+
+4. **Port sync manager**
+   - Copy `cmd/thv-operator/pkg/sync` to `thv-registry-api`
+   - Adapt sync logic to use file-based storage instead of ConfigMap
+   - Implement retry logic (5-minute interval, unlimited attempts)
+   - Remove Kubernetes-specific dependencies
+
+5. **Port filtering logic**
+   - Copy `cmd/thv-operator/pkg/filtering` to `thv-registry-api`
+   - Keep filtering interface and implementation unchanged
+   - Update import paths
+
+6. **Implement automatic sync**
+   - Add `time.Ticker` based periodic sync using `syncPolicy.interval`
+   - Implement hash-based change detection before syncing
+   - Store sync history in file at storage path
+
+7. **Add sync status endpoint**
+   - Implement `GET /api/v1/status` returning sync history
+   - Return format matching proposed JSON schema
+   - Read from sync history file
```

Comment:
I think the shape of the API should come sooner even if it returns placeholders.